From 6845b9d3380d03ae7ae6dddd4ea8018818a4aba0 Mon Sep 17 00:00:00 2001
From: Hariprasad Kelam <hkelam@marvell.com>
Date: Fri, 26 Feb 2021 17:17:40 +0530
Subject: [PATCH 03/17] octeontx2-pf: fix vf ethtool sset count

commit 3b82e4dc5315fc75827843e4e4a7047ca2ab052f from
git@git.assembla.com:cavium/WindRiver.linux.git

Ethtool gets stats info from sset_count and get_strings callbacks.
Current code does not consider complete stats while calculating
sset count for VF devices.This patch fixes the same.

Change-Id: Ibda2ea3b7179bcf0789c3d68a5c2f27a97d4fb6f
Fixes: 1f5499b99("octeontx2-pf: Backport cosmetic upstream changes")
Signed-off-by: Hariprasad Kelam <hkelam@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/46823
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
index a5f299c..0fad3bc 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
@@ -1563,12 +1563,15 @@ static void otx2vf_get_ethtool_stats(struct net_device *netdev,
 static int otx2vf_get_sset_count(struct net_device *netdev, int sset)
 {
 	struct otx2_nic *vf = netdev_priv(netdev);
+	int qstats_count;
 
 	if (sset != ETH_SS_STATS)
 		return -EINVAL;
 
-	return otx2_n_dev_stats +
-	       otx2_n_queue_stats * (vf->hw.rx_queues + vf->hw.tx_queues);
+	qstats_count = otx2_n_queue_stats *
+		       (vf->hw.rx_queues + vf->hw.tx_queues);
+
+	return otx2_n_dev_stats + otx2_n_drv_stats + qstats_count + 1;
 }
 
 static int otx2vf_get_link_ksettings(struct net_device *netdev,
-- 
1.9.1

